### PR TITLE
New module: Add Jboss Cli module (web_infrastructure/jbosscli)

### PR DESCRIPTION
**Pull request by Wolfant**
##### SUMMARY

Is a new module, connect ansible with jboss cli,
jboss-cli.sh is needed on remote host because remote connectios could cause errors when jboss
is running on heavy loads.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME

JBoss-Cli module

##### ANSIBLE VERSION

```
ansible --version
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jan 12 2017, 17:59:37) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```
and

```
ansible 2.1.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```

##### ADDITIONAL INFORMATION

This module help to exec masive jboss cli commands from ansible
  

```

```

### CODE REVIEW

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+    default: localhost:9990
+    description:
+      - JBoss server or domain controller, whit management port
+  verbose:
+    required: false
+    default: false
+    description:
+      - Show the JBoss Cli output, commonly in DMR
+notes:
+  - "jboss-cli.sh need to be runing on client host, and $JAVA_HOME/bin is needeth in Client $PATH"
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: s/needeth/needed/

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+  - ""
+"""
+
+EXAMPLES = """
+# chage scan-interval value, on user wildfy instalation
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: s/instalation/installation/

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+- jboss:
+    command: /subsystem=deployment-scanner/scanner=default:write-attribute(name=scan-interval,value=6000)
+    cli_path: /home/user/wildfly-10.1.0.Final/bin
+
+#  change ExampleDS datasource user-name, on 192.168.20.55:9990 default instalation
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: s/instalation/installation/

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+- jboss:
+    command: /subsystem=datasources/data-source=ExampleDS:write-attribute(name=user-name,value=other)
+    server: 192.168.20.55:9990
+
+# Undeploy the hello world application on wildfy server
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: s/wildfy/wildfly/

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+                    result['stdout'] = re.findall("outcome.+", out)
+        else:
+            result['changed'] = False
+            result['stdout'] = out
+
+    if src:
+        result['changed'] = True
+        result['stdout'] = out
+    if rc != 0:
+        module.fail_json(name='jbosscli', msg=re.findall("failure-description.+", out))
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: Here, you might want to return `out` if `re.findall`  returns `[]`

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+      - JBoss Cli command
+  src:
+    required: false
+    description:
+      - File with list of commands
+  cli_path:
+    required: false
+    default: /usr/share/wildfly/bin
+    description:
+      - The location in the filesystem where jboss-cli.sh is
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: missing 'located.' at end.

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+      - Jboss management user
+  password:
+    required: false
+    description:
+      - Jboss management user password
+  server:
+    required: false
+    default: localhost:9990
+    description:
+      - JBoss server or domain controller, whit management port
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: Possible typo.

File: lib/ansible/modules/web_infrastructure/jbosscli.py (hunk 1 above)
Comment (Akasurde) [reply lib/ansible/modules/web_infrastructure/jbosscli.py]: s/runing/running/

**Review by Akasurde — CHANGES_REQUESTED** [review]

**Comment by sm4rk0** [comment]
Thank you, @Wolfant
That's a valuable contribution.
Did you check what @liquidat did in https://github.com/ansible/ansible-modules-extras/pull/1862 ?
I'm using that module with some [local enhancements](https://gist.github.com/sm4rk0/276b1fdb529fab8d928b7c100fb4f0d8) (changed `command` with `commands` and improved return values).

**Comment by Wolfant** [comment]
@sm4rk0 i see yours and @liquidat, i try to change JBoss Cli output to json but fail when you use --file and the file have 2 or more commands, or when you have "undefined" data, check my with src instead of command

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    default: false
+    description:
+      - Show the JBoss Cli output, commonly in DMR
+notes:
+  - "jboss-cli.sh need to be running on client host, and $JAVA_HOME/bin is needed in Client $PATH"
+  - ""
+"""
+
+EXAMPLES = """
+# chage scan-interval value, on user wildfly installation
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: s/chage/change/

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+# Undeploy the hello world application on wildfly server
+- jboss:
+    command: undeploy hello.war
+    server: "{{ ansible_hostname}}:9990"
+"""
+
+RETURN = '''
+---
+changed:
+    description: if jboss cli command change something
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: s/jboss/JBoss/

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+Error:
+    description: Jboss Cli command error
+    returned: return error if falied
+    type: string
+    sample: '{"outcome" => "success"}'
+
+'''
+
+import os
+import shutil
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: Unused import

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+import time
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: Unused import

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+import re
+import grp
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: Unused import

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+    changed = False
+
+    src = module.params['src']
+    user = module.params['user']
+    password = module.params['password']
+    command = module.params['command']
+    cli_path = module.params['cli_path']
+    server = module.params['server']
+    verbose = module.params['verbose']
+    jsout = None
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: Unused variable.

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+            user=dict(),
+            password=dict(),
+            command=dict(),
+            cli_path=dict(default='/usr/share/wildfly/bin'),
+            server=dict(default='localhost:9990'),
+            verbose=dict(default="False"),
+        ),
+    )
+
+    changed = False
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: Unused variable.

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+    src = module.params['src']
+    user = module.params['user']
+    password = module.params['password']
+    command = module.params['command']
+    cli_path = module.params['cli_path']
+    server = module.params['server']
+    verbose = module.params['verbose']
+    jsout = None
+
+    if command and src:
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: You can use `mutually_exclusive` for `command` and `src`. You can check [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L668) and [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/cloudformation.py#L410)

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+    default: localhost:9990
+    description:
+      - JBoss server or domain controller, with management port
+  verbose:
+    required: false
+    default: false
+    description:
+      - Show the JBoss Cli output, commonly in DMR
+notes:
+  - "jboss-cli.sh need to be running on client host, and $JAVA_HOME/bin is needed in Client $PATH"
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: s/Client/client/

File: lib/ansible/modules/web_infrastructure/jbosscli.py [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+    cli_path = module.params['cli_path']
+    server = module.params['server']
+    verbose = module.params['verbose']
+    jsout = None
+
+    if command and src:
+        module.fail_json(msg="Argument 'src' and 'command' are mutually exclusive")
+
+    if user and not password:
+        module.fail_json(msg="Argument 'user' need 'password' ")
```
Comment (Akasurde) [inline lib/ansible/modules/web_infrastructure/jbosscli.py]: s/need/needs/

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/web_infrastructure/jbosscli.py (hunk 15 above)
Comment (Wolfant) [reply lib/ansible/modules/web_infrastructure/jbosscli.py]: @Akasurde thanks for your advice, and sorry for my English

**Comment by Wolfant** [comment]
ready_for_review

**Comment by smbambling** [comment]
Was going to look at developing a module that used the jboss http api to do something similar.  Any reason you went with using the the jboss-cli.sh script vs hitting the http api end point locally on the host ?

**Comment by wbrefvem** [comment]
@smbambling 
I have been thinking of doing something similar. I've also been looking at providing extra features for [Jolokia](https://jolokia.org/)-enabled instances and have started working on a Jolokia [python client](https://github.com/wbrefvem/python-jolokia) towards that end. My thought was that I would modify the existing jboss module.

**Comment by smbambling** [comment]
@wbrefvem From a quick look it seems the current jboss module just drops a war/ear to be deployed.  If you do end up working on it and want a hand or someone to help test let me know.

**Comment by Wolfant** [comment]
@wbrefvem  the existing jboss module just drop an artifact file and deploy it, do not permit really administration of JBoss.

**Comment by Wolfant** [comment]
@smbambling  thanks, please test it,  It would be very helpful

**Comment by wbrefvem** [comment]
@Wolfant 
@smbambling 
For whomever is interested, we're putting together a JBoss/Wildfly working group to coordinate work around Ansible & JBoss. We need at least two members to get it started. Email me or respond to the [thread](https://groups.google.com/forum/#!topic/ansible-devel/nttHHu0nMxI) on the ansible-devel list.

**Comment by LeoUfimtsev** [comment]
would be nice to have this module available. We have a client who would like to use this functionality.

**Comment by gundalow** [comment]
@LeoUfimtsev Hi, have you tested this module does it work for you?

**Comment by rthaenert** [comment]
I am using this module currently for performing CLI commands on various JBoss instances (EAP 6 as well as EAP 7) and it is working fine.

One minor thing: The error handling is incorrect if `jboss-cli.sh` returns an error code:
```
cause = re.findall("Caused.+", err)
        if cause is not None:
```
re.findall() would return an empty list if there is no match which leads to incomplete error messages (for example if the JRE for the jboss cli is incorrect).

**Comment by hvisage** [comment]
Progress on including this?

Easiest to "test" this locally?

**Pull request closed without merging**